backend/src/database/services/base-service.ts

```diff
@@ -0,0 +1,172 @@
+import type { UUID } from 'digital-fuesim-manv-shared';
+import type {
+    DataSource,
+    DeleteResult,
+    EntityManager,
+    EntityTarget,
+    FindManyOptions,
+    FindOneOptions,
+    FindOptionsWhere,
+} from 'typeorm';
+import { EntityNotFoundError } from 'typeorm';
+import { DatabaseError } from '../database-error';
+import type { BaseEntity } from '../base-entity';
+import type { Creatable, Updatable } from '../dtos';
+
+export abstract class BaseService<TEntity extends BaseEntity> {
+    public constructor(protected readonly dataSource: DataSource) {}
+
+    /**
+     * @param initialObject `undefined` for creating a new object, {@link TEntity} for updating an existing row.
+     */
+    protected abstract createSavableObject<
+        TInitial extends TEntity | undefined
+    >(
+        initialObject: TInitial,
+        dto: TInitial extends TEntity ? Updatable<TEntity> : Creatable<TEntity>,
+        manager: EntityManager
+    ): Promise<TEntity> | TEntity;
+
+    protected abstract readonly entityTarget: EntityTarget<TEntity>;
+
+    public async create(
+        creator: Creatable<TEntity>,
+        entityManager?: EntityManager
+    ): Promise<TEntity> {
+        const create = async (manager: EntityManager) =>
+            manager.save(
+                await this.createSavableObject(
+                    undefined,
+                    // TODO: TypeScript does strange things (@Dassderdie)
+                    creator as any,
+                    manager
+                )
+            );
+        return entityManager
+            ? create(entityManager)
+            : this.dataSource.transaction(create);
+    }
+
+    public async findAll(
+        options?: FindManyOptions<TEntity>,
+        entityManager?: EntityManager
```

Comment:
This code is currently repeated a lot:

```ts
        return entityManager
            ? update(entityManager)
            : this.dataSource.transaction(update);
```

Why not let the caller decide about it (similar as with the selectors in the frontend):

```ts
// without EntityManager:
service.dataSource.transaction(service.assembleFindQuery(...findArgs))
// withEntityManager
assembleFindQuery(...findArgs)(entityManager)
```
(maybe `buildFind` instead of `assemble`)?

Advantage: the API is easier to understand

Disadvantage: a bit more code

Reply:
Another advantage: The programmer (a.k.a. myself) cannot accidentally open multiple transactions as I have to type it out every time.